dist/crawl_links.js

```diff
@@ -0,0 +1,53 @@
+/*
+Collects up to 20 same-origin in-viewport links from img, h[1-6] tags,
+sorted by size from largest to smallest.
+This is the seed for generating a list of URLs to crawl from the given page.
+*/
+const intersectRect = function (r1, r2) {
+    return !(r2.left > r1.right ||
+              r2.right < r1.left ||
+              r2.top > r1.bottom ||
+              r2.bottom < r1.top);
+};
+const sameOrigin = function(uri1, uri2){
+    try {
+        uri1 = new URL(uri1);
+        uri2 = new URL(uri2);
+        return uri1.origin == uri2.origin;
+    } catch(e) {
+    }
+    return false;
+}
+const viewport = {
+    left: 0,
+    top: 0,
+    right: (window.innerWidth || document.documentElement.clientWidth),
+    bottom: (window.innerHeight || document.documentElement.clientHeight)
+};
+let links = {};
+const documentOrigin = window.location.href;
+const elements = document.links;
+for (let e of elements) {
+    try {
+        const rect = e.getBoundingClientRect();
+        const style = window.getComputedStyle(e);
+        if (rect.width > 1 && rect.height > 1 &&
+                e.offsetParent !== null &&
+                style.visibility !== 'hidden' &&
+                style.display !== 'none' &&
+                intersectRect(rect, viewport)) {
+            if (e && e.href != documentOrigin && sameOrigin(e.href, documentOrigin)) {
```

Comment:
Could we also exclude links to the same page, like `#foo`?

I [tested](https://www.webpagetest.org/result/220418_AiDcZG_B85/1/details/#waterfall_view_step1) this on the NYT home page with mobile emulation and got these 5 results:

```
[
    "https://www.nytimes.com/live/2022/04/18/world/ukraine-russia-war-news",
    "https://www.nytimes.com/live/2022/04/18/sports/boston-marathon",
    "https://www.nytimes.com/subscription?campaignId=79FKF",
    "https://www.nytimes.com/#site-content",
    "https://www.nytimes.com/#site-index"
]
```

The last two should probably be omitted.

It also makes me wonder if we should handle mobile differently, since the viewport is much smaller. For example, should we consider nav items hidden behind the hamburger menu?

Desktop [results](https://www.webpagetest.org/result/220418_AiDcB8_B83/1/details/#waterfall_view_step1) for comparison:

```
[
    "https://www.nytimes.com/live/2022/04/18/world/ukraine-russia-war-news",
    "https://www.nytimes.com/2022/04/18/world/europe/ukraine-forces-cluster-munitions.html",
    "https://www.nytimes.com/2022/04/18/world/europe/russian-economy-bleak-assessments.html",
    "https://www.nytimes.com/live/2022/04/18/sports/boston-marathon",
    "https://www.nytimes.com/subscription?campaignId=79FKF",
    "https://www.nytimes.com/international/?action=click&region=Editions&pgtype=Homepage",
    "https://www.nytimes.com/section/t-magazine",
    "https://www.nytimes.com/section/realestate",
    "https://www.nytimes.com/es/",
    "https://www.nytimes.com/section/magazine",
    "https://www.nytimes.com/section/business",
    "https://www.nytimes.com/ca/?action=click&region=Editions&pgtype=Homepage",
    "https://www.nytimes.com/section/science",
    "https://www.nytimes.com/section/opinion",
    "https://www.nytimes.com/section/politics",
    "https://www.nytimes.com/section/health",
    "https://www.nytimes.com/section/sports",
    "https://www.nytimes.com/section/books",
    "https://www.nytimes.com/section/travel",
    "https://www.nytimes.com/section/world"
]
```